File: solutions/ruby/twelve-days/1/twelve_days.rb
class TwelveDays
  DAYS = [
    "first", "second", "third", "fourth", "fifth", "sixth",
    "seventh", "eighth", "ninth", "tenth", "eleventh", "twelfth"
  ]

  GIFTS = [
    "a Partridge in a Pear Tree",
    "two Turtle Doves",
    "three French Hens",
    "four Calling Birds",
    "five Gold Rings",
    "six Geese-a-Laying",
    "seven Swans-a-Swimming",
    "eight Maids-a-Milking",
    "nine Ladies Dancing",
    "ten Lords-a-Leaping",
    "eleven Pipers Piping",
    "twelve Drummers Drumming"
  ]

  def self.song
    (1..12).map { |n| verse(n) }.join("\n\n") + "\n"
  end

  def self.verse(n)
    "On the #{DAYS[n-1]} day of Christmas my true love gave to me: #{gifts_list(n)}."
  end

  def self.gifts_list(n)
    return GIFTS[0] if n == 1
    list = GIFTS[1...n].reverse
    list.join(", ") + ", and " + GIFTS[0]
  end
end

File: solutions/python/twelve-days/1/twelve_days.py
def recite(start_verse, end_verse):
    days = [
        ("first", "a Partridge in a Pear Tree."),
        ("second", "two Turtle Doves, "),
        ("third", "three French Hens, "),
        ("fourth", "four Calling Birds, "),
        ("fifth", "five Gold Rings, "),
        ("sixth", "six Geese-a-Laying, "),
        ("seventh", "seven Swans-a-Swimming, "),
        ("eighth", "eight Maids-a-Milking, "),
        ("ninth", "nine Ladies Dancing, "),
        ("tenth", "ten Lords-a-Leaping, "),
        ("eleventh", "eleven Pipers Piping, "),
        ("twelfth", "twelve Drummers Drumming, ")
    ]

    def get_verse(n):
        day_str, _ = days[n-1]
        verse = f"On the {day_str} day of Christmas my true love gave to me: "
        
        gift_list = []
        for i in range(n-1, -1, -1):
            _, gift = days[i]
            if n > 1 and i == 0:
                gift = "and " + gift
            gift_list.append(gift)
            
        return verse + "".join(gift_list)

    return [get_verse(n) for n in range(start_verse, end_verse + 1)]
